core/src/cluster_slot_state_verifier.rs

```diff
@@ -345,92 +348,120 @@ pub enum ResultingStateChange {
+///    repair
+/// 4) If our local is replayed and matches hash, notify fork choice that we have the correct
+///    version
+fn check_duplicate_confirmed_hash_against_bank_status(
     state_changes: &mut Vec<ResultingStateChange>,
     slot: Slot,
     duplicate_confirmed_hash: Hash,
-    bank_frozen_hash: Hash,
-    is_dead: bool,
+    bank_status: BankStatus,
 ) {
-    if duplicate_confirmed_hash != bank_frozen_hash {
-        if is_dead {
+    match bank_status {
+        BankStatus::Unprocessed => {}
+        BankStatus::Dead => {
             // If the cluster duplicate confirmed some version of this slot, then
             // there's another version of our dead slot
             warn!(
                 "Cluster duplicate confirmed slot {} with hash {}, but we marked slot dead",
                 slot, duplicate_confirmed_hash
             );
-        } else {
+            state_changes.push(ResultingStateChange::RepairDuplicateConfirmedVersion(
+                duplicate_confirmed_hash,
+            ));
+        }
+        BankStatus::Frozen(bank_frozen_hash) if duplicate_confirmed_hash == bank_frozen_hash => {
+            // If the versions match, then add the slot to the candidate
+            // set to account for the case where it was removed earlier
+            // by the `on_duplicate_slot()` handler
+            state_changes.push(ResultingStateChange::DuplicateConfirmedSlotMatchesCluster(
+                bank_frozen_hash,
+            ));
+        }
+        BankStatus::Frozen(bank_frozen_hash) => {
             // The duplicate confirmed slot hash does not match our frozen hash.
             // Modify fork choice rule to exclude our version from being voted
             // on and also repair the correct version
             warn!(
                 "Cluster duplicate confirmed slot {} with hash {}, but our version has hash {}",
                 slot, duplicate_confirmed_hash, bank_frozen_hash
             );
+            state_changes.push(ResultingStateChange::MarkSlotDuplicate(bank_frozen_hash));
+            state_changes.push(ResultingStateChange::RepairDuplicateConfirmedVersion(
+                duplicate_confirmed_hash,
+            ));
         }
-
-        state_changes.push(ResultingStateChange::MarkSlotDuplicate(bank_frozen_hash));
-        state_changes.push(ResultingStateChange::RepairDuplicateConfirmedVersion(
-            duplicate_confirmed_hash,
-        ));
-    } else {
-        // If the versions match, then add the slot to the candidate
-        // set to account for the case where it was removed earlier
-        // by the `on_duplicate_slot()` handler
-        state_changes.push(ResultingStateChange::DuplicateConfirmedSlotMatchesCluster(
-            bank_frozen_hash,
-        ));
     }
 }
 
-fn check_epoch_slots_hash_against_frozen_hash(
+/// Checks the epoch slots hash we observed against our local bank status
+/// Note epoch slots here does not refer to gossip but responses from ancestor_hashes_service
```

Comment:
Can't wait for the follow-up PR to rename all of these references like you suggested in your other comment